### PR TITLE
ENH: stats.ContinuousDistribution.sample: generate QMC samples

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Toward gh-15928
https://stackoverflow.com/questions/79166209/how-to-generate-normally-distributed-sobol-sequences-arrays-in-python/79168966#79168966

#### What does this implement/fix?
This enables generation of quasi-random samples from statistical distributions.

<details>

```python3
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
X = stats.Normal()
qrng = stats.qmc.Sobol(1)
bins = np.linspace(-3.5, 3.5, 31)
plt.hist(X.sample(512, rng=qrng), bins, alpha=0.5, label='quasi-random')
plt.hist(X.sample(512, rng=qrng.rng), bins, alpha=0.5, label='pseudo-random')
plt.title('Histogram of normally-distributed sample')
plt.legend()
plt.show()
```

</details>

image

#### Additional information
How the `shape` parameter is handled:

> If `rng` is an instance of `scipy.stats.QMCEngine` configured to use scrambling and `shape` is not empty, then each slice along the zeroth axis of the result is a "quasi-independent", low-discrepancy sequence; that is, they are distinct sequences that can be treated as statistically independent for most practical purposes. Separate calls to `sample` produce new quasi-independent, low-discrepancy sequences.

So if `len(shape) > 1`, we do this by spawning new RNGs (e.g. `math.prod(shape[1:])` of them if the random variable's parameters are all scalars), instantiatiating a `QMCEngine` with each, and drawing a sample of length `shape[0]` from each of the `QMCEngine`s before passing them through the inverse CDF. Slow, but that's the best we can do with the existing `QMCEngine` interface, I think.

### CODE REVIEW

File: scipy/stats/_qmc.py [hunk 1]
```diff
@@ -8,8 +8,8 @@
 import warnings
 from abc import ABC, abstractmethod
 from functools import partial
+from collections.abc import Callable
```
Comment (mdhaber) [inline scipy/stats/_qmc.py:11]:
```suggestion
```
I guess in `main` this is now imported on line 18.

File: scipy/stats/_probability_distribution.py [hunk 2]
```diff
@@ -123,11 +123,19 @@ def sample(self, shape, *, method, rng):
...
+            operating system. Types other than `numpy.random.Generator` and
+            `scipy.stats.QMCEngine` are passed to `numpy.random.default_rng`
+            to instantiate a ``Generator``.
+
+            If `rng` is an instance of `scipy.stats.QMCEngine` configured to use
+            scrambling and `shape` is not empty, then each slice along the zeroth
+            axis of the result is a "quasi-independent", low-discrepancy sequence;
+            that is, they are distinct sequences that can be treated as statistically
+            independent for most practical purposes. Separate calls to `sample`
+            produce new quasi-independent, low-discrepancy sequences.
```
Comment (tupui) [inline scipy/stats/_probability_distribution.py:138]: As discussed offline. I am not certain where to draw the line and I also think that it is going too deep for a doc in any cases. I would not discuss it and stick to saying sequences using scrambling and a different seed are independent. In many papers such as Art2020 https://arxiv.org/abs/2008.08051 (p8) this is how we do and consider this to be independent. Nevertheless, I sent an email to Sergei to have his take on that.

File: scipy/stats/_probability_distribution.py [hunk 3]
```diff
@@ -123,11 +123,19 @@ def sample(self, shape, *, method, rng):
...
             If the selected `method` is not available, a `NotImplementedError``
             will be raised.
-        rng : `numpy.random.Generator`, optional
-            Pseudorandom number generator state. When `rng` is None, a new
-            `numpy.random.Generator` is created using entropy from the
-            operating system. Types other than `numpy.random.Generator` are
-            passed to `numpy.random.default_rng` to instantiate a `Generator`.
+        rng : `numpy.random.Generator` or `scipy.stats.QMCEngine`, optional
+            Pseudo- or quasi-random number generator state. When `rng` is None,
+            a new `numpy.random.Generator` is created using entropy from the
```
Comment (tupui) [inline scipy/stats/_probability_distribution.py:128]: Just to entertain the thought for 1 second. Did you consider making Halton the default vs MC? There is a slight tradeoff between speed and quality as the shape goes crazy I suppose.

**Review by tupui — COMMENTED** [review]
That would be cool 👍

File: scipy/stats/_probability_distribution.py (hunk 3 above)
Comment (mdhaber) [reply scipy/stats/_probability_distribution.py:128]: We discussed offline that in some cases, such as computational hypothesis testing, you want samples to have independent observations, not low-discrepancy sequences. While QMC is definitely better in some circumstances, it comes with gotchas, so the user should opt in.

File: scipy/stats/_probability_distribution.py (hunk 2 above)
Comment (mdhaber) [reply scipy/stats/_probability_distribution.py:138]: Any word on this?

Comment (tupui) [reply scipy/stats/_probability_distribution.py:138]: Not yet. People are busy with closing the year I suppose 😄 We can keep it as it is for now.

**Comment by mdhaber** [comment]
OK, if CI comes back green otherwise, I'll commit https://github.com/scipy/scipy/pull/21871/#pullrequestreview-2428482088 with `[skip ci]`.

**Comment by tylerjereddy** [comment]
I took a stab at drafting the release note for this (locally only--maybe be good to double check it when I push it up later)

**Comment by mdhaber** [comment]
This is just part of the new infrastructure so it doesn't really need a separate release note, actually.

**Pull request merged**